internal/internal_worker_usage_collector.go

```diff
@@ -0,0 +1,126 @@
+package internal
+
+import (
+	"context"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/uber-go/tally"
+	"go.uber.org/cadence/internal/common/metrics"
+	"go.uber.org/zap"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type (
+	workerUsageCollector struct {
+		workerType   string
+		cooldownTime time.Duration
+		logger       *zap.Logger
+		ctx          context.Context
+		wg           *sync.WaitGroup // graceful stop
+		cancel       context.CancelFunc
+		metricsScope tally.Scope
+	}
+
+	workerUsageCollectorOptions struct {
+		Enabled      bool
+		Cooldown     time.Duration
+		MetricsScope tally.Scope
+		WorkerType   string
+	}
+
+	hardwareUsage struct {
+		NumCPUCores     int
+		CPUPercent      float64
+		NumGoRoutines   int
+		TotalMemory     float64
+		MemoryUsedHeap  float64
+		MemoryUsedStack float64
+	}
+)
+
+func newWorkerUsageCollector(
+	options workerUsageCollectorOptions,
+	logger *zap.Logger,
+) *workerUsageCollector {
+	if !options.Enabled {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &workerUsageCollector{
+		workerType:   options.WorkerType,
+		cooldownTime: options.Cooldown,
+		metricsScope: options.MetricsScope,
+		logger:       logger,
+		ctx:          ctx,
+		cancel:       cancel,
+		wg:           &sync.WaitGroup{},
+	}
+}
+
+func (w *workerUsageCollector) Start() {
+	w.wg.Add(1)
+	go func() {
+		defer func() {
+			if p := recover(); p != nil {
+				w.logger.Error("Unhandled panic in workerUsageCollector.")
+				w.logger.Error(p.(error).Error())
```

Comment:
`p.(error)`is not safe cast. also let's not produce multiple lines here. you can combine these in single error log. see other places where `recover()` is handled

Reply:
Resolved